File: android/app/src/main/java/com/gamespace/app/utils/ShellExecutor.java
package com.gamespace.app.utils;

import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;

public class ShellExecutor {

    public static class CommandResult {
        public final int exitCode;
        public final String stdout;
        public final String stderr;

        public CommandResult(int exitCode, String stdout, String stderr) {
            this.exitCode = exitCode;
            this.stdout = stdout;
            this.stderr = stderr;
        }

        public boolean isSuccess() {
            return exitCode == 0;
        }
    }

    public static CommandResult executeCommand(String command, boolean runAsRoot) {
        Process process = null;
        DataOutputStream os = null;
        BufferedReader isReader = null;
        BufferedReader esReader = null;

        try {
            if (runAsRoot) {
                process = Runtime.getRuntime().exec("su");
                os = new DataOutputStream(process.getOutputStream());
                os.writeBytes(command + "\n");
                os.writeBytes("exit\n");
                os.flush();
            } else {
                process = Runtime.getRuntime().exec(new String[]{"sh", "-c", command});
            }

            StringBuilder stdout = new StringBuilder();
            StringBuilder stderr = new StringBuilder();

            isReader = new BufferedReader(new InputStreamReader(process.getInputStream()));
            esReader = new BufferedReader(new InputStreamReader(process.getErrorStream()));

            String line;
            while ((line = isReader.readLine()) != null) {
                stdout.append(line).append("\n");
            }
            while ((line = esReader.readLine()) != null) {
                stderr.append(line).append("\n");
            }

            int exitCode = process.waitFor();
            return new CommandResult(exitCode, stdout.toString().trim(), stderr.toString().trim());

        } catch (Exception e) {
            return new CommandResult(-1, "", e.getMessage() != null ? e.getMessage() : "Execution exception");
        } finally {
            try {
                if (os != null) os.close();
                if (isReader != null) isReader.close();
                if (esReader != null) esReader.close();
                if (process != null) process.destroy();
            } catch (Exception ignored) {}
        }
    }

    public static boolean isRootAvailable() {
        CommandResult result = executeCommand("id", true);
        return result.isSuccess() && result.stdout.contains("uid=0");
    }
}

File: android/app/src/main/kotlin/com/gamespace/app/utils/ShellExecutor.kt
package com.gamespace.app.utils

import java.io.BufferedReader
import java.io.DataOutputStream
import java.io.InputStreamReader

data class CommandResult(
    val success: Boolean,
    val exitCode: Int,
    val stdout: String,
    val stderr: String
)

object ShellExecutor {

    fun isRootAvailable(): Boolean {
        return try {
            val process = Runtime.getRuntime().exec(arrayOf("su", "-c", "id"))
            val reader = BufferedReader(InputStreamReader(process.inputStream))
            val output = reader.readLine()
            process.waitFor()
            output != null && output.contains("uid=0")
        } catch (e: Exception) {
            false
        }
    }

    /**
     * Sanitizes input to prevent shell command injection.
     */
    private fun sanitizeInput(input: String): String {
        // Allow alphanumeric, underscores, dots, hyphens, commas, and equal signs
        return input.replace(Regex("[^a-zA-Z0-9_\\.\\,-=]"), "")
    }

    fun setSystemPropertyRoot(key: String, value: String): Boolean {
        val cleanKey = sanitizeInput(key)
        val cleanValue = sanitizeInput(value)

        if (cleanKey.isEmpty() || cleanKey.startsWith("ro.")) {
            // Reject empty or read-only ro.* properties
            return false
        }

        val result = executeRootCommand("setprop $cleanKey $cleanValue")
        return result.success && result.exitCode == 0 && result.stderr.isEmpty()
    }

    fun executeRootCommand(command: String): CommandResult {
        return try {
            val process = Runtime.getRuntime().exec("su")
            val os = DataOutputStream(process.outputStream)
            os.writeBytes("$command\n")
            os.writeBytes("exit\n")
            os.flush()

            val stdoutReader = BufferedReader(InputStreamReader(process.inputStream))
            val stderrReader = BufferedReader(InputStreamReader(process.errorStream))

            val stdoutSb = StringBuilder()
            val stderrSb = StringBuilder()

            var line: String?
            while (stdoutReader.readLine().also { line = it } != null) {
                stdoutSb.append(line).append("\n")
            }
            while (stderrReader.readLine().also { line = it } != null) {
                stderrSb.append(line).append("\n")
            }

            val exitCode = process.waitFor()
            val stderrStr = stderrSb.toString().trim()

            CommandResult(
                success = (exitCode == 0 && stderrStr.isEmpty()),
                exitCode = exitCode,
                stdout = stdoutSb.toString().trim(),
                stderr = stderrStr
            )
        } catch (e: Exception) {
            CommandResult(
                success = false,
                exitCode = -1,
                stdout = "",
                stderr = e.localizedMessage ?: "Unknown Error"
            )
        }
    }

    fun getSystemProperty(key: String): String {
        val cleanKey = sanitizeInput(key)
        if (cleanKey.isEmpty()) return ""

        return try {
            val process = Runtime.getRuntime().exec(arrayOf("getprop", cleanKey))
            val reader = BufferedReader(InputStreamReader(process.inputStream))
            val value = reader.readLine()
            process.waitFor()
            value?.trim() ?: ""
        } catch (e: Exception) {
            ""
        }
    }
}
